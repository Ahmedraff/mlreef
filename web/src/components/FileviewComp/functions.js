import FilesApi from 'apis/FilesApi.ts';
import CommitsApi from 'apis/CommitsApi.ts';

const commitsApi = new CommitsApi();

const filesApi = new FilesApi();

const getCommit = (projectId, lastCommitId) => commitsApi
  .getCommitDetails(projectId, lastCommitId);

const getFileAndInformation = (gid, file, branch, commit) => filesApi
  .getFileData(gid, file?.includes('/') ? encodeURIComponent(file) : file, branch || commit)
  .then(async (fData) => {
    const commitInfoDet = await getCommit(gid, fData.last_commit_id);
    return {
      fData,
      commitInfoDet,
    };
  });

export default {
  getFileAndInformation,
};
