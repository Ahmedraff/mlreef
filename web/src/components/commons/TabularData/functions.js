import genUUID from 'uuid/v1';

const parseNumber = (text) => {
  const num = parseFloat(text);

  return Number.isNaN(num) ? text : num;
};

export const transpose = (matrix) => matrix[0].map((_, ci) => matrix.map((row) => row[ci]));

const sanatizeMatrix = (matrix) => {
  const head = matrix[0];
  const tail = matrix[matrix.length - 1];

  const sameLength = head.length === tail.length;
  const lastRowfilled = tail.every((col) => col !== undefined);

  return sameLength && lastRowfilled ? matrix : matrix.slice(0, -1);
};

export const cvsToArray = (text, separator = ',') => {
  const matrix = text.split('\n')
    .map((line) => line.split(separator)
      .map(parseNumber));

  return sanatizeMatrix(matrix);
};

export const jsonToArray = (input) => {
  const items = Array.isArray(input) ? input : Object.values(input);

  if (!items.length) return items;
  // const row = Object.values(input)[0];
  const row = items[0];
  const header = Object.keys(row).map((label) => label);

  // const data = Object.values(input)
  const data = items
    .map((cols) => Object.values(cols).map((val) => val));

  const matrix = [header].concat(data);

  return sanatizeMatrix(matrix);
};

export const arrayToRichData = (matrix, separator = ',') => {
  const header = matrix[0];

  return {
    id: genUUID(),
    meta: {
      separator,
      timestamp: new Date(),
    },
    matrix,
    // data: flat(matrix.map((row, y) => row.map((value, x) => ({ x, y, value }))))
    data: matrix.map((cols, y) => ({
      id: y,
      cols: cols.map((value, x) => ({
        x,
        y,
        value,
        label: header[x],
      })),
    })),
  };
};
