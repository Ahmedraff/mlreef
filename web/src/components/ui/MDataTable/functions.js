const extractField = (row, x) => row.cols.find((col) => col.x === x)?.value;

export const sortDesc = (x) => (rows) => rows.slice()
  .sort((a, b) => extractField(b, x) - extractField(a, x));

export const sortAsc = (x) => (rows) => rows.slice()
  .sort((a, b) => extractField(a, x) - extractField(b, x));

export const round = (decs) => (input) => {
  const decimals = parseInt(decs, 10);

  if (Number.isNaN(decimals)) return input;

  if (decimals > 12) return input;

  const num = parseFloat(input);
  if (Number.isNaN(num)) return input;

  return Math.round(num * (10 ** decimals)) / (10 ** decimals);
};
