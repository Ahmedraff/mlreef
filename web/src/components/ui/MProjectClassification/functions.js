import { ALGORITHM, DATA_OPERATION, MODEL, OPERATION, VISUALIZATION } from "dataTypes";

export const getSearchable = (feSearchableTypes) => {
  switch (feSearchableTypes) {
    case DATA_OPERATION:
      return OPERATION;
    case MODEL:
      return ALGORITHM;
    default:
      return VISUALIZATION;
  }
}