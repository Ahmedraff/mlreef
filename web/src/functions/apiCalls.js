import { POLL_TIMEOUT } from 'apiConfig';
import { PROJECT_TYPES } from 'domain/project/projectTypes';
import CommitsApi from '../apis/CommitsApi.ts';

const commitsApi = new CommitsApi();

export const getCommits = (projectId, commitBranch) => commitsApi.getCommits(projectId, commitBranch, '', 1);

export const getCommitDetails = (
  projectId, 
  commitId
) => commitsApi.getCommitDetails(projectId, commitId);

export const getFileDifferences = async (projectId, diff, previousCommitId, lastCommitId) => {
  let previousVersionFile;
  let nextVersionFile;
  let imageFileSize;
  if (!diff.new_file) {
    await commitsApi.getFileDataInCertainCommit(
      projectId,
      encodeURIComponent(
        diff.old_path,
      ), previousCommitId,
    )
      .then((res) => {
        previousVersionFile = res.imageArrayBuffer;
        imageFileSize = res.imageFileSize;
      });
  }
  if (!diff.deleted_file) {
    await commitsApi.getFileDataInCertainCommit(
      projectId,
      encodeURIComponent(
        diff.old_path,
      ), lastCommitId,
    )
      .then((res) => {
        nextVersionFile = res.imageArrayBuffer;
        imageFileSize = res.imageFileSize;
      });
  }

  return { previousVersionFile, nextVersionFile, imageFileSize };
};

/**
 * Suscribe to a real time (polling) communication.
 *
 * @param {Object} options
 * @param {Number[integer]} options.timeout interval in milliseconds (default 1000).
 * @param {Function} action the function to be called.
 * @param {any} args the parameter for the function.
 *
 * @return {Function} the unsuscribe function.
 */
export const suscribeRT = (options = {}) => (action, args) => {
  const {
    timeout,
  } = options;

  let timeoutId = null;

  const executeTimedAction = () => {
    action(args);

    timeoutId = setTimeout(executeTimedAction, timeout || POLL_TIMEOUT);
  };

  executeTimedAction();

  return () => {
    clearTimeout(timeoutId);
  };
};

// this returns an error if code is bigger than 400
// added an extra guard to avoid failing by bad json parsing
export const handleResponse = (res) => res
  .json()
  .then((body) => {
    if (!res.ok) {
      const error = new Error();
      error.name = res.statusText;
      error.message = body.message;
      return Promise.reject(error);
    }
    return res.status !== 204 ? body : res;
  });

export const inspect = (res) => console.info(res) || res;

export const onlyDataProject = (project) => project.searchableType === PROJECT_TYPES.DATA;

export const onlyCodeProject = (project) => project.searchableType === PROJECT_TYPES.CODE;

export const handlePagination = ({ content }) => ([...content]);
