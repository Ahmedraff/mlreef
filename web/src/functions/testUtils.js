import { createStore, applyMiddleware } from 'redux';

import rootReducer from '../reducers/index';
import { middlewares } from '../configureStore';

// eslint-disable-next-line import/prefer-default-export
export const storeFactory = (initialState) => {
  const createStoreWithMiddleware = applyMiddleware(...middlewares)(createStore);
  return createStoreWithMiddleware(rootReducer, initialState);
};
