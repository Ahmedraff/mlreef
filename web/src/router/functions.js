/**
 * Compose like: f.g x = g(f(x))
 *
 * @example compose (fn4, fn3, fn2, fn1) (arg) === fn4(fn3(fn2(fn1(arg))))
 *
 * @param {Array[Function]} fns sequence of functions.
 * @param arg input parameter.
 *
 * @return far left funcion return.
 */
export const c = (...fns) => (arg) => fns.reverse().reduce((acc, fn) => fn(acc), arg);

/**
 * Return the recorded route that matches name.
 */
export const findRouteInfo = (routes) => ({ name }) => routes.find((route) => route.name === name);

/**
 * Replace placeholders by parameters. e.g. /:user/messages -> /nick/messages.
 *
 * @param {Object(Route)} to implicit.
 * @param {Object} params e.g. key-val object, from the to object.
 * @param {Object(Route)} record the route from routes.
 *
 * @return {Object(Route)} route with path modified.
 */
export const replaceParams = (to) => (record = {}) => {
  if (to.params) {
    const path = Object.entries(to.params)
      .reduce((final, [k, v]) => final.replace(new RegExp(`:${k}`), v), record.path || '');

    return { ...record, path };
  }

  return record;
};

/**
 * Add the hash part. e.g. #last-menu-item
 *
 * @param {Object(Route)} to implicit.
 * @param {String} hash e.g. last-menu-item, from the to object.
 * @param {Object(Route)} record the route from routes.
 *
 * @return {Object(Route)} route with path modified.
 */
export const addHash = (to) => (record) => !to.hash ? record
  : { ...record, path: `${record.path}#${to.hash}` };

export const makeQuery = (query) => Object.entries(query)
  .map(([k, v]) => `${k}=${v}`)
  .join('&');

/**
 * Add the query search. e.g. ?user=nickname&priority=high
 *
 * @param {Object(Route)} to implicit.
 * @param {Object} query e.g. key-val object, from the to object.
 * @param {Object(Route)} record the route from routes.
 *
 * @return {Object(Route)} route with path modified.
 */
export const addQuery = (to) => (record) => !to.query ? record
  : { ...record, path: `${record.path}?${makeQuery(to.query)}` };

export const unpack = ({ path }) => path;

export const resolveRoute = (routes, to) => c(
  unpack, addQuery(to), addHash(to), replaceParams(to), findRouteInfo(routes),
)(to);
