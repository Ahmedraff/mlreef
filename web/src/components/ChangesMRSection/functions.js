import { Base64 } from 'js-base64';
import Base64ToArrayBuffer from 'base64-arraybuffer';
import { getFileDifferences } from '../../functions/apiCalls';
import CommitsApi from 'apis/CommitsApi';
import { isImageFormat } from 'functions/dataParserHelpers';

const commitsApi = new CommitsApi();

export const fetchFileDiffInfo = (projectId, aheadCommits) => Promise.all(aheadCommits.map(async (commit, index) => {
  const { id: commitId } = commit;
  const commitDiffs = await commitsApi.getCommitDiff(projectId, commitId, 0, false);
  return await Promise.all(commitDiffs?.body?.map(async (commitDiff) => {
    const {
      previousVersionFile,
      nextVersionFile,
      imageFileSize,
    } = await getFileDifferences(projectId, commitDiff, commit.parent_ids[0], commit.id);
    let previousVersionFileParsed = previousVersionFile;
    let nextVersionFileParsed = nextVersionFile;
    let sizeDeleted = 0;
    let sizeAdded = 0;
    const fileName = commitDiff.old_path.split('/').slice(-1)[0];
    if (previousVersionFile && !isImageFormat(fileName)) {
      previousVersionFileParsed = Base64.decode(
        Base64ToArrayBuffer.encode(previousVersionFile),
      );
      sizeDeleted = previousVersionFile.byteLength;
    }
    if (nextVersionFile && !isImageFormat(fileName)) {
      nextVersionFileParsed = Base64.decode(Base64ToArrayBuffer.encode(nextVersionFile));
      sizeAdded = nextVersionFile.byteLength;
    }
    const path = commitDiff.new_path
      ? commitDiff.new_path
      : commitDiff.old_path;
    const fileExtension = path.split('.').pop().toLowerCase();
    return {
      fileName,
      fileExtension,
      previousVersionFileParsed,
      nextVersionFileParsed,
      id: `${commit.id} ${index}`,
      sizeDeleted,
      sizeAdded,
      imageFileSize,
    };
  }));
}));
