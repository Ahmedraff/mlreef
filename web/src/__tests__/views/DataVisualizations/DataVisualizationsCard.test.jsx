import React from 'react';
import renderer from 'react-test-renderer';
import { BrowserRouter as Router } from 'react-router-dom';
import { Provider } from 'react-redux';
import { storeFactory } from 'functions/testUtils';
import { mount } from 'enzyme';
import DataVisualizationCard from 'components/views/DataVisualization/dataVisualizationCard';
import classifiedMockedVisualizations from './testData';

const fireModal = jest.fn();

const fetchVisualizations = jest.fn();

test('assert that snapshot matches', () => {
  const store = storeFactory({});
  const snapShot = renderer.create(
    <Provider store={store}>
      <Router>
        <DataVisualizationCard
          classification={classifiedMockedVisualizations[0]}
          namespace="mlreef"
          slug="project-slug"
          key={classifiedMockedVisualizations[0].status}
          fireModal={fireModal}
          callback={fetchVisualizations}
        />
      </Router>
    </Provider>,
  ).toJSON();
  expect(snapShot).toMatchSnapshot();
});

const setup = (index = 0) => {
  const store = storeFactory({});
  const testWrapper = mount(
    <Provider store={store}>
      <Router>
        <DataVisualizationCard
          classification={classifiedMockedVisualizations[index]}
          namespace="mlreef"
          slug="project-slug"
          key={classifiedMockedVisualizations[index].status}
          fireModal={fireModal}
          callback={fetchVisualizations}
        />
      </Router>
    </Provider>,
  );

  return testWrapper;
};

describe('assert basic UI text elements are rendered', () => {
  let wrapper;

  test('pipeline info is displayed', () => {
    wrapper = setup();
    expect(wrapper.find('img').prop('alt')).toEqual('running');
    expect(wrapper.find('.title-div').text()).toBe('In progress');
    expect(wrapper.find('Link').find('a').text()).toBe('fresh-seal-18012021225907-2');
    expect(wrapper.find('a').at(1).text()).toBe('mlreef');
  });

  test('assert that firemodal is called with running', () => {
    wrapper = setup(0);
    wrapper.find('button.btn-danger').simulate('click');
    expect(fireModal).toHaveBeenCalled();
    expect(fireModal.mock.calls[0][0].title.includes('Abort')).toBeTruthy();
  });

  test('assert that firemodal is called with running', () => {
    wrapper = setup(3);
    wrapper.find('button.btn-danger').simulate('click');
    expect(fireModal).toHaveBeenCalled();
    expect(fireModal.mock.calls[0][0].title.includes('Delete')).toBeTruthy();
  });

  afterEach(() => {
    fireModal.mockClear();
  });
});
