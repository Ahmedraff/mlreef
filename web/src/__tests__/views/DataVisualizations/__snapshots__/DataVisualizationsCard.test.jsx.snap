// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`assert that snapshot matches 1`] = `
<div
  className="pipeline-card"
>
  <div
    className="header"
  >
    <div
      className="title-div"
    >
      <p>
        <b>
          In progress
        </b>
      </p>
    </div>
  </div>
  <div
    className="data-visualization-card-content"
    data-key="data-visualization/fresh-seal-18012021225907-2"
  >
    <img
      alt="running"
      height="30"
      src={
        Object {
          "canInstrument": true,
          "createTransformer": [Function],
          "getCacheKey": [Function],
          "process": [Function],
        }
      }
      width="30"
    />
    <div
      className="general-information ml-2"
    >
      <a
        href="/mlreef/project-slug/-/visualizations/b3b05d6b-d3ae-4c58-8bf9-ed7a47f34566"
        onClick={[Function]}
      >
        <b>
          fresh-seal-18012021225907-2
        </b>
      </a>
      <p
        className="m-0 mt-1"
      >
        Created by
         
        <a
          href="/mlreef"
        >
          <b>
            mlreef
          </b>
        </a>
         
        3 month(s)
        Â  ago
      </p>
    </div>
    <div
      className="auth-wrapper ownership-required group-role-required"
      title="Please login"
    >
      <div
        className="auth-wrapper-cover"
        onClick={[Function]}
      />
      <div
        className="auth-wrapper-wrapped"
      >
        <button
          className="btn btn-danger border-solid my-auto"
          onClick={[Function]}
          style={
            Object {
              "width": "max-content",
            }
          }
          type="button"
        >
          Abort
        </button>
      </div>
    </div>
  </div>
</div>
`;
